ui-src/NVRApplication.js

```diff
@@ -0,0 +1,290 @@
+ *
+ * @param  {String} timeZone Timezone name
+ */
+function newTimeZone(timeZone) {
+  timeFormatter = new TimeFormatter(timeFmt, timeZone);
+  timeFormatter90k = new TimeStamp90kFormatter(timeZone);
+}
+
+/**
+ * Globally set a new time format for the app.
+ *
+ * @param  {String} format Time format specification
+ */
+function newTimeFormat(format) {
+  timeFormatter = new TimeFormatter(format, timeFormatter.tz);
+}
+
+/**
+ * Event handler for clicking on a video.
+ *
+ * A 'dialog' object is attached to the body of the dom and it
+ * properly initialized with the corrcet src url.
+ *
+ * @param  {NVRSettings} nvrSettingsView NVRSettingsView in effect
+ * @param  {object} camera Object for the camera
+ * @param  {object} range Range Object
+ * @param  {object} recording Recording object
+ * @return {void}
+ */
+function onSelectVideo(nvrSettingsView, camera, range, recording) {
+  console.log('Recording clicked: ', recording);
+  const trimmedRange = recording.range90k(nvrSettingsView.trim ? range : null);
+  console.log('Forming video url');
+  const url = api.videoPlayUrl(
+    camera.uuid,
+    recording,
+    trimmedRange,
+    nvrSettingsView.timeStampTrack
+  );
+  console.log('Video url: ' + url);
+  const video = $('<video controls preload="auto" autoplay="true" />');
+  const dialog = $('<div class="playdialog" />').append(video);
+  console.log('have dialog');
+  $('body').append(dialog);
+  console.log('appended dialog');
+
+  let [formattedStart, formattedEnd] = timeFormatter90k.formatSameDayShortened(
+    trimmedRange.startTime90k,
+    trimmedRange.endTime90k
+  );
+  console.log('range: ' + formattedStart + '-' + formattedEnd);
+  dialog.dialog({
+    title: camera.shortName + ', ' + formattedStart + ' to ' + formattedEnd,
+    width: recording.videoSampleEntryWidth / 4,
+    close: function() {
+      dialog.remove();
+    },
+  });
+  // Now that dialog is up, set the src so video starts
+  console.log('Video setting rc: ', url);
+  video.attr('src', url);
+}
+
+/**
+ * Fetch camera view data for a given date/time range.
+ *
+ * @param  {Range90k} selectedRange Desired time range
+ * @param  {Number} videoLength Desired length of video segments
+ */
+function fetch(selectedRange, videoLength) {
+  if (selectedRange.startTime90k === null) {
+    return;
+  }
+  console.log(
+    'Fetching> ' +
+      selectedRange.formatTimeStamp90k(selectedRange.startTime90k) +
+      ' to ' +
+      selectedRange.formatTimeStamp90k(selectedRange.endTime90k)
+  );
+  for (let cameraView of cameraViews) {
+    let url = api.recordingsUrl(
+      cameraView.camera.uuid,
+      selectedRange.startTime90k,
+      selectedRange.endTime90k,
+      videoLength
+    );
+    if (url === cameraView.recordingsUrl) {
+      /*
+       * @TODO: Can this actually happen?
+       */
+      continue; // already in progress, nothing to do.
+    }
+    if (cameraView.recordingsReq !== null) {
+      /*
+       * @TODO: Aborting here does not see right.
```

Comment:
Why do you say that? Let's say this happens:

1. user selects a date range; a fetch starts
2. user selects a different date range while that fetch is still going

isn't the right thing to do simply to abort the fetch from step 1 and replace it with a different one? Isn't that basically what this is doing?

I'm wondering if there's an easy way to artificially delay the response from the server to test it out.

Reply:
ping - still wondering about this

Reply:
I guess that is a viable scenario. The need to abort seems to then stem from two sources: (1) there is only a single request variable so to make the new request, you'd have to abort the current one, (2) aborting is more efficient than simply making another request, and handling the result of the prior one.
I suppose even if you'd do that, results might be out of sequence. So, I'll remove the TODO.